apm-sniffer/apm-sdk-plugin/micronaut-plugins/micronaut-http-client-plugin/src/main/java/org/apache/skywalking/apm/plugin/micronaut/http/client/ClientExchangeInterceptor.java

```diff
@@ -0,0 +1,121 @@
+
+        MutableHttpRequest<?> request = (MutableHttpRequest<?>) allArguments[0];
+
+        final ContextCarrier contextCarrier = new ContextCarrier();
+        String requestMethod = request.getMethod().name();
+        URL peerURL = new URL((String) objInst.getSkyWalkingDynamicField());
+        AbstractSpan span = ContextManager.createExitSpan(requestMethod + ":" + request.getPath(), contextCarrier, peerURL.getHost() + ":" + peerURL.getPort());
+        span.setComponent(ComponentsDefine.MICRONAUT_CLIENT);
+        Tags.HTTP.METHOD.set(span, requestMethod);
+        String url = objInst.getSkyWalkingDynamicField() + request.getPath();
+        Tags.URL.set(span, url);
+        SpanLayer.asHttp(span);
+        CarrierItem next = contextCarrier.items();
+        while (next.hasNext()) {
+            next = next.next();
+            request.header(next.getHeadKey(), next.getHeadValue());
+        }
+        span.prepareForAsync();
+        ContextManager.stopSpan(span);
+        request.setAttribute(ASYNC_SPAN_KEY, span);
+        if (MicronautHttpClientPluginConfig.Plugin.MicronautHttpClient.COLLECT_HTTP_PARAMS) {
+            collectHttpParam(request, span);
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret) throws Throwable {
+
+        Publisher<HttpResponse<?>> retPublisher = (Publisher<HttpResponse<?>>) ret;
+        MutableHttpRequest<?> request = (MutableHttpRequest<?>) allArguments[0];
+        return Flux.from(retPublisher)
+                .doOnError(ex ->
+                        request.getAttribute(ASYNC_SPAN_KEY)
+                                .map(span -> (AbstractSpan) span)
+                                .ifPresent(span -> span.errorOccurred().log(ex))
+                )
+                .doOnNext(resp -> {
+                    request.getAttribute(ASYNC_SPAN_KEY)
+                            .map(span -> (AbstractSpan) span)
+                            .ifPresent(span -> {
+                                Tags.HTTP_RESPONSE_STATUS_CODE.set(span, resp.code());
+                                if (resp.code() >= 400) {
+                                    span.errorOccurred();
+                                }
+                            });
+                }).doFinally(e -> {
+                    request.getAttribute(ASYNC_SPAN_KEY)
+                            .map(span -> (AbstractSpan) span)
+                            .ifPresent(span -> {
+                                // Active HTTP parameter collection automatically in the profiling context.
+                                if (!MicronautHttpClientPluginConfig.Plugin.MicronautHttpClient.COLLECT_HTTP_PARAMS && span.isProfiling()) {
+                                    collectHttpParam(request, span);
+                                }
+                                span.asyncFinish();
```

Comment:
This should be placed in a final, in case an exception occurs in the param deserialization and cause context leaks.